src/org/loklak/harvester/strategy/BlockingKaizenHarvester.java

```diff
@@ -0,0 +1,61 @@
+package org.loklak.harvester.strategy;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.loklak.data.DAO;
+
+public class BlockingKaizenHarvester extends KaizenHarvester {
+
+    /**
+     * This class uses blocking queue to store queries which need to be processed.
+     */
+    public BlockingKaizenHarvester() {
+        super(new KaizenQueries() {
+            private int maxSize = DAO.getConfig("harvester.kaizen.queries_limit", 500);
+            private BlockingQueue<String> queries = new ArrayBlockingQueue<>(maxSize);
+            private int blockingTimeout = DAO.getConfig("harvester.blocking_kaizen.block_time", 120);
+
+            @Override
+            public boolean addQuery(String query) {
+                if (this.queries.contains(query)) {
+                    return false;
+                }
+                try {
+                    /*
+                    * We need to associate a timeout with the addition process as there are chances (though little)
```

Comment:
The [`BlockingQueue`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingQueue.html) handles all that internally.

Also, I didn't get that how timers can be a problem. The internal mutex will allow only one of the waiting queries will get added.